### PR TITLE
Import chrome extension list

### PR DESCRIPTION
**Pull request by simonhong**
BraveExternalProcessImporterHost will fetch available extensions list
and ask to install them.
Other importing options are done via other utility process but
importing extension should be done in browser process.

NOTE: this commits has many cleanup changes around importing area
that uses deprecated patching style.

Resolves brave/brave-browser#7974

import_chrome_extensions

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
`npm run test brave_unit_tests -- --filter=ChromeImporterUtilsTest`

Manual test step
1. Launch Chrome and install some extensions from webstore
2. Close Chrome and launch Brave
3. Import from chrome and check extensions option
4. After seeing import done dialog, wait and check extensions are installed.
## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by simonhong** [comment]
@rebron Can you check the new option name?

**Comment by rebron** [comment]
@simonhong That looks good. Thanks.

File: patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch [hunk 1]
```diff
@@ -15,7 +15,9 @@ index 4192db5cdfb0ff0600460ac20b7151e4794a9d40..94033e6abd64c1cefdc7ae9a721f8bda
 +        !(this.getPref('import_dialog_ledger').value &&
 +          this.selected_.ledger) &&
 +        !(this.getPref('import_dialog_windows').value &&
-+          this.selected_.windows);
++          this.selected_.windows) &&
++        !(this.getPref('import_dialog_extensions').value &&
++          this.selected_.extensions);
```
Comment (simonhong) [inline patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch]: @petemill please advice if you have any idea about avoiding this patch.

File: common/pref_names.cc [hunk 2]
```diff
@@ -78,6 +78,11 @@ const char kRemoteDebuggingEnabled[] = "brave.remote_debugging_enabled";
 const char kAutocompleteEnabled[] = "brave.autocomplete_enabled";
 const char kBraveDarkMode[] = "brave.dark_mode";
 const char kOtherBookmarksMigrated[] = "brave.other_bookmarks_migrated";
+const char kImportDialogCookies[] = "import_dialog_cookies";
```
Comment (bsclifton) [inline common/pref_names.cc]: nice cleanup here 😄👍

File: patches/chrome-browser-importer-external_process_importer_client.cc.patch [hunk 3]
```diff
@@ -1,17 +1,12 @@
...
    localized_strings.try_emplace(
        IDS_BOOKMARK_BAR_FOLDER_NAME,
        l10n_util::GetStringUTF8(IDS_BOOKMARK_BAR_FOLDER_NAME));
-+  localized_strings.try_emplace(
-+      IDS_IMPORT_FROM_BRAVE,
-+      l10n_util::GetStringUTF8(IDS_IMPORT_FROM_BRAVE));
-+  localized_strings.try_emplace(
-+      IDS_BOOKMARK_GROUP_FROM_BRAVE,
-+      l10n_util::GetStringUTF8(IDS_BOOKMARK_GROUP_FROM_BRAVE));
++  BRAVE_START
```
Comment (bsclifton) [inline patches/chrome-browser-importer-external_process_importer_client.cc.patch]: 😎👌

**Comment by simonhong** [comment]
@bsclifton Thanks for testing and review!
Yes, I also thought that many patches in this area could be deleted after we deprecate muon import feature :)

**Comment by simonhong** [comment]
All CI passed.

**Comment by bsclifton** [comment]
cc: @bridiver for patch review

File: patches/chrome-browser-importer-external_process_importer_host.cc.patch [hunk 4]
```diff
@@ -0,0 +1,20 @@
+diff --git a/chrome/browser/importer/external_process_importer_host.cc b/chrome/browser/importer/external_process_importer_host.cc
+index 9087093838f249ca73cc9fc82d54d5b44e9bb5f2..22e47149317aa6f7a48661d58156f176a2be33e8 100644
+--- a/chrome/browser/importer/external_process_importer_host.cc
++++ b/chrome/browser/importer/external_process_importer_host.cc
+@@ -60,6 +60,7 @@ void ExternalProcessImporterHost::StartImportSettings(
+     return;
+   }
+ 
++  BRAVE_START_IMPORT_SETTINGS
```
Comment (bridiver) [inline patches/chrome-browser-importer-external_process_importer_host.cc.patch]: why are we patching here when we have a subclass?

File: patches/chrome-browser-importer-external_process_importer_host.cc.patch [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+   CheckForLoadedModels(items);
+ 
+   LaunchImportIfReady();
+@@ -113,6 +114,7 @@ void ExternalProcessImporterHost::LaunchImportIfReady() {
+   client_ = new ExternalProcessImporterClient(
+       weak_ptr_factory_.GetWeakPtr(), source_profile_, items_, bridge);
+   client_->Start();
++  BRAVE_LAUNCH_IMPORT_IF_READY
```
Comment (bridiver) [inline patches/chrome-browser-importer-external_process_importer_host.cc.patch]: same as above and I definitely think we should not be patching for p3a, we can find another place to do that

File: chromium_src/chrome/browser/importer/external_process_importer_host.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_CHROME_BROWSER_IMPORTER_EXTERNAL_PROCESS_IMPORTER_HOST_H_
+#define BRAVE_CHROMIUM_SRC_CHROME_BROWSER_IMPORTER_EXTERNAL_PROCESS_IMPORTER_HOST_H_
+
+#define BRAVE_EXTERNAL_PROCESS_IMPORTER_HOST_H \
+ private: \
+  friend class BraveExternalProcessImporterHost; \
+  virtual bool CheckForChromeOrBraveLock(); \
```
Comment (bridiver) [inline chromium_src/chrome/browser/importer/external_process_importer_host.h]: this doesn't need to be added to the header, you can cast when you call it

File: patches/chrome-browser-importer-external_process_importer_client.cc.patch (hunk 3 above)
Comment (bridiver) [reply patches/chrome-browser-importer-external_process_importer_client.cc.patch]: can't we do this in `StartImport` instead of patching?

File: patches/chrome-browser-importer-external_process_importer_host.cc.patch (hunk 4 above)
Comment (simonhong) [reply patches/chrome-browser-importer-external_process_importer_host.cc.patch]: I assume that this was added here because it needs early return in the middle of this method like above upstream condition.

File: patches/chrome-browser-importer-external_process_importer_host.cc.patch (hunk 5 above)
Comment (simonhong) [reply patches/chrome-browser-importer-external_process_importer_host.cc.patch]: How about leaving this as a f/u? This goes  beyond of the scope of this PR.
This PR just does cleanup current patching around importer area.

File: patches/chrome-browser-importer-external_process_importer_client.cc.patch (hunk 3 above)
Comment (simonhong) [reply patches/chrome-browser-importer-external_process_importer_client.cc.patch]: Yes, I think it's more complicated and maybe it also needs another patching.
To do that, we have to subclass ProfileImportProxy::StartImport() and it should be replaced with BraveProfileImportProxy in `chrome/common/importer/profile_import.mojom.h`.
So, I think one line patching here is better. WDYT?

File: patches/chrome-browser-importer-external_process_importer_host.cc.patch (hunk 5 above)
Comment (simonhong) [reply patches/chrome-browser-importer-external_process_importer_host.cc.patch]: fixed by https://github.com/brave/brave-core/pull/4672/commits/8c029feb736f5134caa3f6a3f398ab404b7d4cfd

File: browser/ui/webui/settings/brave_settings_import_data_handler_base.cc [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    uint16_t imported_items) {
+  ImportDataHandler::StartImport(source_profile, imported_items);
+
+  type_ = source_profile.importer_type;
+}
+
+void BraveImportDataHandlerBase::ImportEnded() {
+  ImportDataHandler::ImportEnded();
+
+  if (!import_did_succeed_)
```
Comment (simonhong) [inline browser/ui/webui/settings/brave_settings_import_data_handler_base.cc]: @iefremov should we always report regardless of import result?

File: patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch (hunk 1 above)
Comment (bridiver) [reply patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch]: I'm fairly certain there's a better way to do this

File: patches/chrome-browser-importer-external_process_importer_host.cc.patch (hunk 4 above)
Comment (bridiver) [reply patches/chrome-browser-importer-external_process_importer_host.cc.patch]:
```#define CheckForFirefoxLock \
  UnusedMethod() { return false; } \
  virtual bool CheckForFirefoxLock``` and then override in the subclass

File: patches/chrome-browser-importer-external_process_importer_client.cc.patch (hunk 3 above)
Comment (bridiver) [reply patches/chrome-browser-importer-external_process_importer_client.cc.patch]: Why would you need to patch the mojo stuff? Just subclass ProfileImportImpl and chromium_src override in chrome/utility/services.cc

Comment (simonhong) [reply patches/chrome-browser-importer-external_process_importer_client.cc.patch]: Ah, ok. I didn't think to override in utility process side.
and I found `ProfileImportImpl` also has some patches that could be cleanup together :)

File: browser/importer/brave_external_process_importer_host.h [hunk 8]
```diff
@@ -1,32 +1,33 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (iefremov) [inline browser/importer/brave_external_process_importer_host.h]: 2020? :)

File: common/importer/chrome_importer_utils.h [hunk 9]
```diff
@@ -1,19 +1,17 @@
...
-
+#include <string>
 #include <vector>
 
-namespace base {
-class DictionaryValue;
-class FilePath;
-class ListValue;
-}
+#include "base/values.h"
```
Comment (iefremov) [inline common/importer/chrome_importer_utils.h]: can be replaced with forward declaration?

File: common/importer/chrome_importer_utils.h [hunk 10]
```diff
@@ -1,19 +1,17 @@
...
+#include "base/files/file_path.h"
```
Comment (iefremov) [inline common/importer/chrome_importer_utils.h]: move this one to .cc &

File: patches/chrome-utility-importer-external_process_importer_bridge.h.patch [hunk 11]
```diff
@@ -1,12 +1,12 @@
...
-index 58db1e3c50c057bdf49795d4042ce633e7a80890..e828f0ff701ab473c038c78d4b98ad40e548b6a4 100644
+index 58db1e3c50c057bdf49795d4042ce633e7a80890..c4b61125d6ba75769f8348831b723571b8fbb40c 100644
 --- a/chrome/utility/importer/external_process_importer_bridge.h
 +++ b/chrome/utility/importer/external_process_importer_bridge.h
-@@ -73,6 +73,7 @@ class ExternalProcessImporterBridge : public ImporterBridge {
+@@ -71,6 +71,7 @@ class ExternalProcessImporterBridge : public ImporterBridge {
+   base::string16 GetLocalizedString(int message_id) override;
+   // End ImporterBridge implementation.
  
++  BRAVE_EXTERNAL_PROCESS_IMPORTER_BRIDGE_H
```
Comment (iefremov) [inline patches/chrome-utility-importer-external_process_importer_bridge.h.patch]: why is this change needed? Friends can be declared in any part of the class (private/public/whatever), so I just don't get why a macro and corresponding chromium_src is needed

File: common/importer/importer_constants.h [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMMON_IMPORTER_IMPORTER_CONSTANTS_H_
+#define BRAVE_COMMON_IMPORTER_IMPORTER_CONSTANTS_H_
+
+// Pref file that holds installed extension list.
+extern const char kChromeExtensionsPreferencesFile[];
+extern const char kChromeExtensionsListPath[];
```
Comment (iefremov) [inline common/importer/importer_constants.h]: you can just use `constexpr char kChromeExtensionsPreferencesFile[] = "...";` and avoid adding .cc

File: chromium_src/chrome/browser/importer/external_process_importer_host.h [hunk 13]
```diff
@@ -0,0 +1,19 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_CHROME_BROWSER_IMPORTER_EXTERNAL_PROCESS_IMPORTER_HOST_H_
+#define BRAVE_CHROMIUM_SRC_CHROME_BROWSER_IMPORTER_EXTERNAL_PROCESS_IMPORTER_HOST_H_
+
+#define BRAVE_EXTERNAL_PROCESS_IMPORTER_HOST_H \
+ private: \
```
Comment (iefremov) [inline chromium_src/chrome/browser/importer/external_process_importer_host.h]: imo we just don't need to add an access modifier for friend directives

File: browser/ui/webui/settings/brave_settings_import_data_handler_base.cc (hunk 7 above)
Comment (iefremov) [reply browser/ui/webui/settings/brave_settings_import_data_handler_base.cc]: I think it's not v important given that probably import should be mostly successful. But we can verify it with @davidtemkin . In general, I think we should just preserve the original behaviour

File: patches/chrome-utility-importer-external_process_importer_bridge.h.patch (hunk 11 above)
Comment (simonhong) [reply patches/chrome-utility-importer-external_process_importer_bridge.h.patch]: Its subclass `BraveExternalProcessImporterBridge` accesses private member of `ExternalProcessImporterBridge`.

File: browser/importer/brave_external_process_importer_host.h (hunk 8 above)
Comment (simonhong) [reply browser/importer/brave_external_process_importer_host.h]: I did this intentionally because this file was created last year.

Comment (simonhong) [reply browser/importer/brave_external_process_importer_host.h]: Hmm, should we use 2020 if we change license sentence?

File: browser/resources/settings/brave_settings_overrides.js [hunk 14]
```diff
@@ -398,6 +398,38 @@ BravePatching.RegisterPolymerTemplateModifications({
       version.innerHTML = '<a id="release-notes" target="_blank" href="https://brave.com/latest/">' + version.innerHTML + '</a>'
     }
   },
+  'settings-import-data-dialog': (templateContent) => {
+    let checkBoxesParent = templateContent.querySelector('#browserSelect').parentElement
```
Comment (petemill) [inline browser/resources/settings/brave_settings_overrides.js]: 🤔 ✅

File: patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch (hunk 1 above)
Comment (petemill) [reply patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch]: Yes I'm trying to see if we can override this via behavior

Comment (petemill) [reply patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch]: I think we can remove this whole patch and extend `prefsChanged_` via a behavior. 

In brave_settings_overrides:
```js
BravePatching.RegisterPolymerComponentBehaviors({
  'settings-import-data-dialog': [{
  registered: function () {
    const oldPrefsChanged = this.prefsChanged_
    if (!oldPrefsChanged) {
      console.error('[Brave Settings Overrides] cannot find prefsChanged_ on ImportDataDialog')
      return
    }
    this.prefsChanged_ = function () {
      if (typeof this.noImportDataTypeSelected_ !== 'boolean') {
        console.error('[Brave Settings Overrides] cannot find noImportDataTypeSelected_ on ImportDataDialog')
        return
      }
      oldPrefsChanged.apply(this)
      if (this.selected_ == undefined || this.prefs == undefined) {
        return;
      }
      this.noImportDataTypeSelected_ = this.noImportDataTypeSelected_ &&
        !(this.getPref('import_dialog_cookies').value &&
        this.selected_.cookies) &&
        !(this.getPref('import_dialog_stats').value &&
          this.selected_.stats) &&
        !(this.getPref('import_dialog_ledger').value &&
          this.selected_.ledger) &&
        !(this.getPref('import_dialog_windows').value &&
          this.selected_.windows) &&
        !(this.getPref('import_dialog_extensions').value &&
          this.selected_.extensions)
    }
  }
}]
})
```

Another option would be to instead override `shouldDisableImport_` which would prevent having to copy the early return, but that changes the logic from a push to a pull, and I'm not sure if the notification will work.

Comment (simonhong) [reply patches/chrome-browser-resources-settings-people_page-import_data_dialog.js.patch]: @petemill Thanks! this works well. :thu

File: browser/importer/brave_external_process_importer_host.h (hunk 8 above)
Comment (bsclifton) [reply browser/importer/brave_external_process_importer_host.h]: I think it's OK as-is 😄

**Comment by simonhong** [comment]
@bridiver @petemill PTAL again :)

**Comment by simonhong** [comment]
Rebased on C81.

File: patches/chrome-utility-importer-external_process_importer_bridge.h.patch (hunk 11 above)
Comment (bridiver) [reply patches/chrome-utility-importer-external_process_importer_bridge.h.patch]: this also brings the patch inline with current patching standards

File: common/importer/chrome_importer_utils.cc [hunk 15]
```diff
@@ -66,13 +99,49 @@ bool ChromeImporterCanImport(const base::FilePath& profile,
...
 }
+
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+std::vector<std::string> GetImportableListFromChromeExtensionsList(
+    const base::Value& extensions_list) {
+  DCHECK(extensions_list.is_dict());
+
+  std::vector<std::string> extensions;
+  for (const auto& item : extensions_list.DictItems()) {
+    // Only import if type is extension, it's came from webstore and it's not
```
Comment (bridiver) [inline common/importer/chrome_importer_utils.cc]: we should also be checking to make sure it is enabled, but what about data in places like `chrome.storage` ? Are we expecting that users import extensions with all current settings or are we automatically installing whatever they had before?

Comment (simonhong) [reply common/importer/chrome_importer_utils.cc]: This PR targets just getting all explicitly installed extension list.
WDYT? @bsclifton @rebron

Comment (bsclifton) [reply common/importer/chrome_importer_utils.cc]: Yup- let's do the storage as a follow up. We can create a separate issue for this 👍

Comment (bsclifton) [reply common/importer/chrome_importer_utils.cc]: @simonhong @bridiver created https://github.com/brave/brave-browser/issues/8847 for that functionality! As-is, this is good to go 👍

Comment (simonhong) [reply common/importer/chrome_importer_utils.cc]: Yeah, it's good to have as a f/u :) Thanks for creating that issue.

**Comment by simonhong** [comment]
CI has failed test `PerfPredictorTabHelperTest.NewNavigationStoresSavings` on linux only.
and other platforms are ok. I think this test is flaky and not related with this PR. Merged.
